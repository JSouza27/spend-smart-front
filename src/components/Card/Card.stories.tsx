import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ThemeProvider } from 'styled-components';

import Card from '.';
import theme from '../../styles/theme';

export default {
  title: 'Card',
  component: Card
} as ComponentMeta<typeof Card>;

export const Basic: ComponentStory<typeof Card> = (args) => (
  <ThemeProvider theme={theme}>
    <Card {...args} />
  </ThemeProvider>
);
