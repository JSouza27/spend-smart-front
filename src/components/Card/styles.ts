import styled, { css } from 'styled-components';
import { CardProps } from '.';

type ContainerProps = {
  colorByType: string;
} & Pick<CardProps, 'type'>;

export const Wrapper = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.neutral_0};
    border: 1px solid ${theme.colors.neutral_400};
    border-radius: 0.8rem;
    box-shadow: 0px 4px 8px rgba(16, 24, 64, 0.08);
    flex: 1 1 auto;
    padding: ${theme.spacings.xsmall};
    max-width: 60rem;

    @media (min-width: 767px) {
      padding: ${theme.spacings.small};
    }
  `}
`;

export const Container = styled.div<ContainerProps>`
  ${({ theme, colorByType, type }) => css`
    align-items: center;
    color: ${colorByType};
    display: flex;
    font-size: clamp(2rem, 0.87vw + 1.76rem, 2.8rem);
    font-weight: ${theme.font.bold};
    justify-content: space-between;

    .icon {
      color: ${colorByType};
      transform: ${type === 'expense' && 'scale(-1) rotate(270deg)'};
    }

    span::before {
      content: 'R$';
      font-size: ${theme.font.sizes.small};
      margin-right: 1rem;
    }
  `}
`;
