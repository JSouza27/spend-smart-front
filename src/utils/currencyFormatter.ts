export class CurrencyFormatter {
  private static language = 'pt-BR';
  private static currency = 'BRL';

  static formatter(num: number): string {
    return (num ?? 0).toLocaleString(this.language, {
      currency: this.currency
    });
  }

  static unformat(num: string): number {
    const nums = num.replace(/[^0-9]+/g, '');
    const i = nums.length - 2;

    return Number(`${nums.substring(0, i)}.${nums.substring(i)}`);
  }
}
