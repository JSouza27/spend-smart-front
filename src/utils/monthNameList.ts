export const monthNameList = [
  'janeiro',
  'fevereiro',
  'mar√ßo',
  'abril',
  'maio',
  'junho',
  'julho',
  'agosto',
  'setembro',
  'outubro',
  'novembro',
  'dezembro'
];
