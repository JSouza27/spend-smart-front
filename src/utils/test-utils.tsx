import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../styles/theme';
import { FormProvider, useForm } from 'react-hook-form';

const AllTheProviders = ({ children }: any) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>{children}</FormProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: any
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
