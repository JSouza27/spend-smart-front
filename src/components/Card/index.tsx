import { ArrowTrending, Savings } from '@styled-icons/fluentui-system-filled';

import theme from '../../styles/theme';
import { CurrencyFormatter } from '../../utils/currencyFormatter';
import Heading from '../Heading';
import * as S from './styles';

export type CardProps = {
  title: string;
  type: 'expense' | 'income' | 'balance';
  value: number;
};

export default function Card({ title, type, value }: CardProps) {
  const balanceResult = (value: number): string => {
    if (value > 0) {
      return theme.colors.green_400;
    } else if (value < 0) {
      return theme.colors.red_400;
    } else {
      return theme.colors.neutral_800;
    }
  };

  const colorByType = (): string => {
    if (type === 'income') return theme.colors.green_400;
    if (type === 'expense') return theme.colors.red_400;
    if (type === 'balance') return balanceResult(value);

    return theme.colors.neutral_800;
  };

  return (
    <S.Wrapper>
      <Heading level={3} size="small">
        {title}
      </Heading>

      <S.Container colorByType={colorByType()} type={type}>
        <span>{CurrencyFormatter.formatter(value).replace('R$', '')}</span>
        {type === 'balance' ? (
          <Savings className="icon" size={38} title="icon" />
        ) : (
          <ArrowTrending className="icon" size={38} title="icon" />
        )}
      </S.Container>
    </S.Wrapper>
  );
}
