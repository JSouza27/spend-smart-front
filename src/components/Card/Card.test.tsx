import { render, screen } from 'utils/test-utils';

import Card from '.';

describe('<Card />', () => {
  it('should render the Card', () => {
    const { container } = render(
      <Card title="Total de Receitas" type="expense" value={17354.61} />
    );

    const title = screen.getByRole('heading', {
      name: /total de receitas/i
    });
    const number = screen.getByText(/17\.354,61/i);
    const icon = screen.getByTitle('icon');

    expect(title).toBeInTheDocument();
    expect(number).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
