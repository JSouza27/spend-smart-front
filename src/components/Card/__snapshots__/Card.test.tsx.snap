// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render the Card 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  color: #101840;
  font-size: 1.4rem;
  font-weight: 600;
  line-height: 4rem;
}

.c0 {
  background-color: #FFFFFF;
  border: 1px solid #D8DAE5;
  border-radius: 0.8rem;
  box-shadow: 0px 4px 8px rgba(16,24,64,0.08);
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  padding: 1.6rem;
  max-width: 40rem;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #52BD94;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: clamp(2rem,0.87vw + 1.76rem,2.8rem);
  font-weight: 700;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 .icon {
  color: #52BD94;
}

.c2 span::before {
  content: 'R$';
  font-size: 1.4rem;
  margin-right: 1rem;
}

@media (min-width:767px) {
  .c0 {
    padding: 2.4rem;
  }
}

<div
  class="c0"
>
  <h3
    class="c1"
  >
    Total de Receitas
  </h3>
  <div
    class="c2"
    type="expense"
  >
    <span>
      17.354,61
    </span>
    <svg
      class="c3 icon"
      fill="currentColor"
      focusable="false"
      height="38"
      role="img"
      viewBox="0 0 24 24"
      width="38"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        icon
      </title>
      <path
        d="M14 5.5h7a1 1 0 0 1 1 .88v7.12a1 1 0 0 1-2 .12V8.9l-7.3 7.3a1 1 0 0 1-1.31.08l-.1-.08L9 13.9l-5.28 5.3a1 1 0 0 1-1.5-1.32l.08-.1 6-6a1 1 0 0 1 1.32-.08l.1.08L12 14.1l6.58-6.59H14a1 1 0 0 1-.99-.88V6.5a1 1 0 0 1 .88-1H21h-7Z"
      />
    </svg>
  </div>
</div>
`;
